### PR TITLE
fix(use-click-away-listener): rewrite ClickAwayWrapper component as a custom hook

### LINKED ISSUE #5245: [ ClickAwayWrapper ] the prop `targets[0].current` issue
### Current behaviour

The newly added wrapper `ClickAwayWrapper` has an issue in TS. When getting the reference to the `targets[]` the first render is `null`

### Expected behaviour

There should not be any of errors

### CodeSandbox or Storybook URL

https://codesandbox.io/s/silly-goodall-81rv2x

### JIRA Ticket (Sage Only)

_No response_

### Suggested Solution

_No response_

### Carbon Version

107.1.7

### Design Tokens Version

_No response_

### What browsers are you seeing the problem on?

Firefox, Chrome, Safari, Microsoft Edge

### What Operating System are you seeing the problem on?

MacOS

### Anything else we should know?

_No response_

### Confidentiality

- [X] I confirm there is no confidential or commercially sensitive information included.

### PR DESCRIPTION
**Pull request by edleeks87**
fix #5245

### Proposed behaviour

Rewrites `ClickAwayWrapper` component as `usClickAwayListener` custom hook
Removes `targets` from `useEffect` dependency to prevent unnecessarily triggering the `handleClickAway` callback
Adds hook to `MultiActionButton` and `SplitButton`

### Current behaviour

`ClickAwayWrapper` component exists

### Checklist

- [x] Commits follow our style guide
- [x] Related issues linked in commit messages if required
- [x] Typescript `d.ts` file added or updated if required

#### QA

- [ ] Tested in CodeSandbox/storybook
- [ ] Add new Cypress test coverage if required
- [ ] Carbon implementation matches Design System/designs
- [ ] UI Tests GitHub check reviewed if required

### Testing instructions

Updates can be tested in storybook for `PopoverContainer`, `Menu`, `Date`, `DateRange`, `MultiActionButton` and `SplitButton`

The following CodeSandbox is an example of the broken behaviour.
You can see the new behaviour by looking at the version in the comment by `codesandbox[bot]`.

### CODE REVIEW

File: src/__internal__/click-away-wrapper/click-away-wrapper.component.tsx [hunk 1]
```diff
@@ -14,11 +19,13 @@ const ClickAwayWrapper = ({
   eventTypeId = "click",
   targets,
 }: ClickAwayWrapperProps) => {
+  const targetsRef = useRef(targets);
+
```
Comment (mkrds) [inline src/__internal__/click-away-wrapper/click-away-wrapper.component.tsx:23]:
```suggestion
targetsRef.current = targets
```

Comment (mkrds) [reply src/__internal__/click-away-wrapper/click-away-wrapper.component.tsx:23]: The `targetsRef` needs to be updated on each rerender to make this pattern work.

Comment (edleeks87) [reply src/__internal__/click-away-wrapper/click-away-wrapper.component.tsx:23]: It isn't needed for any of the components that currently implement it, the only scenario I found that required it if the `targets` array were dynamically updated (adding/ removing/ changing the ref that points to an element). I'll add it in as we'll likely need it when/ if we add ref forwarding to them or if any new components do dynamically alter the `targets`.

https://codesandbox.io/s/beautiful-mayer-kkflin?file=/src/index.js

**Review by DipperTheDan — DISMISSED** [review]

File: src/hooks/__internal__/useClickAwayListener/useClickAwayListener.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        eventTypeId="mousedown"
+      />
+    );
+    fireEvent.mouseDown(document);
+    expect(handleClickAway).toHaveBeenCalled();
+  });
+
+  it("does not call handleClickAway when mousedown is inside of wrapper element", () => {
+    const handleClickAway = jest.fn();
+    render(<MockComponent handleClickAway={handleClickAway} />);
```
Comment (mkrds) [inline src/hooks/__internal__/useClickAwayListener/useClickAwayListener.spec.tsx:47]:
```suggestion
    render(<MockComponent handleClickAway={handleClickAway} eventTypeId="mousedown" />);
```

**Review by DipperTheDan — COMMENTED** [review]
Nothing more to add than what @mkrds has already commented.

File: src/components/multi-action-button/multi-action-button.component.tsx [hunk 3]
```diff
@@ -244,6 +231,8 @@ export const MultiActionButton = ({
     </Popover>
   );
 
+  useClickAwayListener([ref], hideButtons);
```
[the remark this answers is not in the document]
Comment (edleeks87) [reply src/components/multi-action-button/multi-action-button.component.tsx:234]: I've confirmed that it isn't needed as the onClick passed to the children elements of the container call the `hideButtons` function

**Review by mkrds — APPROVED** [review]

**Pull request merged**